projects/flow.groups-2D-random-walk/src/project.py

```diff
@@ -0,0 +1,211 @@
+@RandomWalkProject.operation
+def compute_radius_of_gyration(job):
+    """Compute the radius of gyration for a random walk."""
+    with job.data:
+        positions = job.data["positions"][:]
+    avg_position = positions.mean(axis=0)
+    deviation = positions - avg_position
+    job.doc.radius_of_gyration = np.sqrt(
+        np.sum(deviation * deviation) / deviation.shape[0]
+    )
+    job.doc.average_position = avg_position
+
+
+@base
+@RandomWalkProject.pre.after(simulate)
+@RandomWalkProject.post.true("end_to_end")
+@RandomWalkProject.operation
+def compute_end_to_end_distance(job):
+    """Compute the radius of gyration for a random walk."""
+    with job.data:
+        positions = job.data["positions"]
+        job.doc["end_to_end"] = np.linalg.norm(positions[-1] - positions[0])
+
+
+@base
+@RandomWalkProject.pre.after(simulate)
+@RandomWalkProject.post(lambda job: "squared_distance" in job.data)
+@RandomWalkProject.operation
+def compute_squared_distance(job):
+    """Compute the squared distance for a random walk."""
+    with job.data:
+        positions = job.data.positions[:]
+    job.data.squared_distance = np.sum(positions * positions, axis=1)
+
+
+# This operation happens after computing the msd so it isn't in base.
+# Also we use pre as a filter for jobs ensuring this job only ever runs on the zeroth
+# replica for a given standard deviation.
+@RandomWalkProject.pre.true("msd_analyzed")
+@RandomWalkProject.pre(lambda job: job.sp.replica == 0)
+@RandomWalkProject.post.isfile("msd.png")
+@RandomWalkProject.operation
+def plot_msd(job):
+    """Plot the MSD for all std."""
+    with job.data:
+        msd = job.data.msd[:]
+    fig, ax = plt.subplots()
+    ax.plot(msd)
+    ax.set_title(f"MSD for {job.sp.standard_deviation} std")
+    ax.set_xlabel("x")
+    ax.set_ylabel("MSD")
+    # Only save figure to the first replica
+    fig.savefig(job.fn("msd.png"))
+    fig.close()
+
+
+# Aggregate operations on individual standard deviations across replicas
+
+# Create aggregator that combines all replicas with a single standard deviation
+std_aggregator = flow.aggregator.groupby("standard_deviation")
+
+
+# Define all aggregate groups
+agg_plot = RandomWalkProject.make_group(
+    "aggregate-plot", group_aggregator=std_aggregator
+)
+agg_analyze_and_plot = RandomWalkProject.make_group(
+    "post-processing", group_aggregator=std_aggregator
+)
+
+
+@agg_analyze_and_plot
+@std_aggregator
+@RandomWalkProject.pre(all_simulated)
+@RandomWalkProject.post.true("msd_analyzed")
+@RandomWalkProject.operation
+def compute_mean_squared_distance(*jobs):
+    """Compute and store the mean squared displacement for all std."""
+    squared_distance_iterator = generate_stores(jobs, "squared_distance")
+    squared_distance = next(squared_distance_iterator)
+    msd = np.copy(squared_distance)
+    for squared_distance in squared_distance_iterator:
+        msd += squared_distance
+    msd /= len(jobs)
```

Comment:
I think this is equivalent and easier to understand.
```suggestion
    msd = np.zeros(jobs[0].doc.run_steps)
    for squared_distance in generate_stores(jobs, "squared_distance"):
        msd += squared_distance
    msd /= len(jobs)
```

Reply:
I think this is equivalent and easier to understand.
```suggestion
    msd = np.zeros(jobs[0].doc.run_steps + 1)
    for squared_distance in generate_stores(jobs, "squared_distance"):
        msd += squared_distance
    msd /= len(jobs)
```